lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2444,19 +2444,120 @@ namespace {
+      else {
+        // If this is a nested namespace, we need to find its extension decl
+        // context and then use that to find the parent enum. It's important
+        // that we add this to the parent enum (in the "__ObjC" module) and not
+        // the extension.
+        if (auto parentNS = dyn_cast<clang::NamespaceDecl>(decl->getParent())) {
+          auto parent = Impl.importDecl(parentNS, getVersion());
+          // Sometimes when the parent namespace is imported, this namespace
+          // also gets imported. If that's the case, then the parent namespace
+          // will be an enum (because it was able to be fully imported) in which
+          // case we need to bail here.
+          auto cachedResult =
+              Impl.ImportedDecls.find({decl->getCanonicalDecl(), getVersion()});
+          if (cachedResult != Impl.ImportedDecls.end())
+            return cachedResult->second;
+          dc = cast<ExtensionDecl>(parent)
+                   ->getExtendedType()
+                   ->getEnumOrBoundGenericEnum();
+        } else {
+          // Otherwise, this is a non-namesapce decl, so it's owned by its
+          // module in both C++ and Swift meaning we can proceed normally.
+          dc = extensionDC;
+        }
+      }
+      auto *enumDecl = Impl.createDeclWithClangNode<EnumDecl>(
           decl, AccessLevel::Public, loc,
           importedName.getDeclName().getBaseIdentifier(),
           Impl.importSourceLoc(decl->getLocation()), None, nullptr, dc);
-      enumDecl->setMemberLoader(&Impl, 0);
+      if (isa<clang::NamespaceDecl>(decl->getParent()))
+        cast<EnumDecl>(dc)->addMember(enumDecl);
+
+      // We are creating an extension, so put it at the top level. This is done
+      // after creating the enum, though, because we may need the "correctly"
+      // nested decl context above when creating the enum.
+      while (extensionDC->getParent() &&
+             extensionDC->getContextKind() != DeclContextKind::FileUnit)
+        extensionDC = extensionDC->getParent();
+
+      auto *extension = ExtensionDecl::create(Impl.SwiftContext, loc, nullptr,
+                                              {}, extensionDC, nullptr, decl);
+      Impl.SwiftContext.evaluator.cacheOutput(ExtendedTypeRequest{extension},
+                                              enumDecl->getDeclaredType());
+      Impl.SwiftContext.evaluator.cacheOutput(ExtendedNominalRequest{extension},
+                                              std::move(enumDecl));
+      // Keep track of what members we've already added so we don't add the same
+      // member twice. Note: we can't use "ImportedDecls" for this because we
+      // might import a decl that we don't add (for example, if it was a
+      // parameter to another decl).
+      SmallPtrSet<Decl *, 16> addedMembers;
+      for (auto redecl : decl->redecls()) {
+        // This will be re-set as the enum after we return from
+        // "VisitNamespaceDecl."
```

Comment:
A call to `importDecl` will "cache" the imported decl (the result from the visitor). We set `Impl.ImportedDecls[redecl]` here so that subdecls will have the correct decl context, but once this function returns an enum, `importDecl`'s implementation will "cache" or reset `Impl.ImportedDecls[decl]` with the imported enum.

Reply:
Thanks! Please add this information to the comment.

"This will be reset" -> "`importDecl` will reset the imported decl to the EnumDecl..."